mono/metadata/icall.c

```diff
@@ -7299,13 +7299,27 @@ ves_icall_System_Environment_Exit (int result)
 	if (!mono_runtime_try_shutdown ())
 		mono_thread_exit ();
 
-	/* Suspend all managed threads since the runtime is going away */
-	mono_thread_suspend_all_other_threads ();
-
-	mono_runtime_quit_internal ();
+	/*
+	 * Suspend all managed threads since the runtime is going away and if
+	 * that succeeds tear down the runtime. Otherwise we are just going to
+	 * exit().
+	 */
+	if (mono_thread_suspend_all_other_threads ())
+		mono_runtime_quit_internal ();
 
 	/* we may need to do some cleanup here... */
+
+#ifdef HOST_WIN32
+	/*
+	 * Wine msvcrt's exit() calls CorExitProcess() if mscoree is loaded.
```

Comment:
Wine's `CorExitProcess` is supposed to detect that mono has begun shutting down and call `ExitProcess` in that case, but it seems detecting that relies on `mono_profiler_set_runtime_shutdown_begin_callback`, which I'm guessing is part of `mono_runtime_quit_internal`.

Looking at it again now, it probably would've been better to check `mono_runtime_is_shutting_down`, but since we need to maintain compatibility with current Wine, I wouldn't want Wine Mono (at least for 9.x) to depend on that. So I think we'll want to keep this for now as a Wine Mono specific hack.

Reply:
I've split this into a separate commit so you don't have to apply it over to Mono repo and keep it local to Wine-Mono.